### PR TITLE
MecanumDriveOdometry to handle odometry estimation of Mecanum wheeled models

### PR DESCRIPTION
**Pull request by danilogsch**
# 🎉 New feature

Closes [#683](https://github.com/gazebosim/gz-sim/pull/683)

## Summary
 Math Class to handle odometry estimation of Mecanum wheeled models. 
Observation: Aproximation when the robot is given both linear, lateral and angular velocity commands. Exact pose integration or Runge Kuffa second order integration happens if one the velocity straight commands (linear or lateral) is 0.

## Test it
run mecanum_drive.sdf and listen the odometry and tf topics.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

### CODE REVIEW

File: include/ignition/math/MecanumDriveOdometry.hh [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2019 Open Source Robotics Foundation
```
Comment (ahcorde) [inline include/ignition/math/MecanumDriveOdometry.hh:2]:
```suggestion
 * Copyright (C) 2022 Open Source Robotics Foundation
```

File: include/ignition/math/MecanumDriveOdometry.hh [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    // Inline bracket to help doxygen filtering.
+    inline namespace IGNITION_MATH_VERSION_NAMESPACE {
+    //
+    // Forward declarations.
+    class MecanumDriveOdometryPrivate;
+
+    /** \class MecanumDriveOdometry MecanumDriveOdometry.hh \
+     * ignition/math/MecanumDriveOdometry.hh
+     **/
```
Comment (ahcorde) [inline include/ignition/math/MecanumDriveOdometry.hh:41]:
```suggestion
    /// \class MecanumDriveOdometry MecanumDriveOdometry.hh \
    /// ignition/math/MecanumDriveOdometry.hh
    ///
```

File: include/ignition/math/MecanumDriveOdometry.hh [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    /// A vehicle with a heading of zero degrees has a local
+    /// reference frame according to the diagram below.
+    ///
+    ///       Y
+    ///       ^
+    ///       |
+    ///       |
+    ///       O--->X(forward)
+    ///
+
```
Comment (ahcorde) [inline include/ignition/math/MecanumDriveOdometry.hh:54]:
```suggestion
```

File: src/MecanumDriveOdometry.cc [hunk 4]
```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (C) 2019 Open Source Robotics Foundation
```
Comment (ahcorde) [inline src/MecanumDriveOdometry.cc:2]:
```suggestion
 * Copyright (C) 222 Open Source Robotics Foundation
```

File: src/MecanumDriveOdometry.cc [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+// https://research.ijcaonline.org/volume113/number3/pxc3901586.pdf
+
+class ignition::math::MecanumDriveOdometryPrivate
+{
+  /// \brief Integrates the pose.
+  /// \param[in] _linear Linear velocity.
+  /// \param[in] _lateral Lateral velocity.
+  /// \param[in] _angular Angular velocity.
+  public: void IntegrateExact(double _linear, double _lateral, double _angular);
+  public: void IntegrateRungeKutta2(double _linear, double _lateral, double _angular);
```
Comment (ahcorde) [inline src/MecanumDriveOdometry.cc:36]: add docs

**Review by ahcorde — CHANGES_REQUESTED** [review]
Please check 
 - linters https://github.com/gazebosim/gz-math/runs/8020312206?check_suite_focus=true
 - Add tests

We are also adding python interfaces to new classes, if you don't have time to do it, please open an issue about adding python interfaces.

**Comment by danilogsch** [comment]
depends on [#1664](https://github.com/gazebosim/gz-sim/pull/1664) of gz-sim

File: src/MecanumDriveOdometry.cc (hunk 5 above)
Comment (danilogsch) [reply src/MecanumDriveOdometry.cc:41]:
```suggestion
  public: void IntegrateRungeKutta2(double _linear, double _lateral, double _angular);
```
```suggestion
   /// \brief Integrates the poseusing second order Runge-Kuffa aproximation.
  /// \param[in] _linear Linear velocity.
  /// \param[in] _lateral Lateral velocity.
  /// \param[in] _angular Angular velocity.
  public: void IntegrateRungeKutta2(double _linear, double _lateral, double _angular);
```

File: src/MecanumDriveOdometry.cc (hunk 4 above)
Comment (danilogsch) [reply src/MecanumDriveOdometry.cc]: commited the 222 by mistake

**Review by ahcorde — CHANGES_REQUESTED** [review]
linters https://github.com/gazebosim/gz-math/actions/runs/3534393561/jobs/5931205749

```
  /github/workspace/include/ignition/math/MecanumDriveOdometry.hh:81:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:36:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  /github/workspace/src/MecanumDriveOdometry.cc:41:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:144:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:152:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:153:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:154:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:155:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:176:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:177:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:180:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:181:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:182:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:206:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:272:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:285:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:286:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/src/MecanumDriveOdometry.cc:287:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  Total errors found: 18
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
linters https://github.com/gazebosim/gz-math/actions/runs/3583238905/jobs/6034063390

**Comment by azeey** [comment]
@danilogsch Do you think you will have time to fix the linter issues and address @ahcorde's  [feedback](https://github.com/gazebosim/gz-math/pull/486#pullrequestreview-1085732557) ?

File: include/ignition/math/MecanumDriveOdometry.hh [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+      /// \brief Updates the odometry class with latest wheels and
+      /// steerings position
+      /// \param[in] _frontLeftPos Left wheel position in radians.
+      /// \param[in] _frontRightPos Right wheel postion in radians.
+      /// \param[in] _backLeftPos Left wheel position in radians.
+      /// \param[in] _backRightPos Right wheel postion in radians.
+      /// \param[in] _time Current time point.
+      /// \return True if the odometry is actually updated.
+      public: bool Update(const Angle &_frontLeftPos, const Angle &_frontRightPos, const Angle &_backLeftPos, const Angle &_backRightPos,
+                          const clock::time_point &_time);
```
Comment (ahcorde) [inline include/ignition/math/MecanumDriveOdometry.hh:82]:
```suggestion
      public: bool Update(const Angle &_frontLeftPos, const Angle &_frontRightPos, 
                          const Angle &_backLeftPos, const Angle &_backRightPos,
                          const clock::time_point &_time);
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by danilogsch** [comment]
hey, sorry that I haven't reached here in a while... Is there anything else you need me to do here?

**Review by danilogsch — COMMENTED** [review]
I missed this ones

**Comment by Bi0T1N** [comment]
Looks like linter errors are resolved, only Ubuntu Bionic lists
```
/github/workspace/include/ignition/math/MecanumDriveOdometry.hh:81:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
/github/workspace/include/ignition/math/MecanumDriveOdometry.hh:81:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```
but this could probably be neglected since Focal/Jammy don't complain.

However, you have to add tests - probably similar to the ones in [`DiffDriveOdometry_TEST.cc`](https://github.com/gazebosim/gz-math/blob/gz-math7/src/DiffDriveOdometry_TEST.cc).  
Also your [last commit](https://github.com/gazebosim/gz-math/pull/486/commits/889b05c2aaf12c732ded4bdb4834e6f42d0dde65) isn't signed but one option *could* be to amend, sign and force-pushing the new commit to your branch.

**Comment by danilogsch** [comment]
@Bi0T1N , thanks for the insight... I had no idea on how to deal with unsigned commits.
Anyway, I have time this week and I will work on the tests.

**Comment by muttistefano** [comment]
Hi, is there anything I can do to help here?
I have been waiting for this to be merged for a while, so if I can help, I would be glad.
Thanks

**Comment by danilogsch** [comment]
> Hi, is there anything I can do to help here? I have been waiting for this to be merged for a while, so if I can help, I would be glad. Thanks

Hey, thanks for asking... could you help me with the tests? I've been using Garden in my research, I did not have time to look into the [DiffDriveOdometry_TEST.cc ](https://github.com/gazebosim/gz-math/blob/gz-math7/src/DiffDriveOdometry_TEST.cc) and to mimic that just yet

**Comment by muttistefano** [comment]
ok,i can take a look into that.
I have a question regarding the pull, I see it targets ign-math6, this means that after that we will have to do another one into the gz-math7 branch ? 
Thanks

**Comment by Bi0T1N** [comment]
> I have a question regarding the pull, I see it targets ign-math6, this means that after that we will have to do another one into the gz-math7 branch ? Thanks

Yes, as `ign-math6` is only for Citadel and Fortress. The merge into `gz-math7` should be quite simple as it doesn't touch many other files but it should be rebased on the current HEAD due to the `ignition` to `gz` renaming.

**Comment by danilogsch** [comment]
> ok,i can take a look into that. I have a question regarding the pull, I see it targets ign-math6, this means that after that we will have to do another one into the gz-math7 branch ? Thanks

I made this two branches when Garden was released:
[gz-sim7](https://github.com/danilogsch/gz-sim/tree/gz-sim7-mwodom)
[gz-math7](https://github.com/danilogsch/gz-math/tree/gz-math7-mwodom)

basic the same as this PR, feel free to use it

File: include/gz/math/MecanumDriveOdometry.hh [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      /// \param[in] _wheelSeparation Distance between left and right wheels.
+      /// \param[in] _wheelBase Distance between front and back wheels.
+      /// \param[in] _leftWheelRadius Radius of the left wheel.
+      /// \param[in] _rightWheelRadius Radius of the right wheel.
+      public: void SetWheelParams(double _wheelSeparation, double _wheelBase,
+                      double _leftWheelRadius, double _rightWheelRadius);
+
+      /// \brief Set the velocity rolling window size.
+      /// \param[in] _size The Velocity rolling window size.
+      public: void SetVelocityRollingWindowSize(size_t _size);
```
Comment (azeey) [inline include/gz/math/MecanumDriveOdometry.hh:117]: For a follow-up PR, we should also add getters for the parameters set by these API.

**Comment by mjcarroll** [comment]
I just addressed the linting issues here and pulled in @muttistefano's test from #537

File: include/gz/math/MecanumDriveOdometry.hh [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    /// properties and wheel speeds for a Mecanum-drive vehicle.
+    ///
+    /// A vehicle with a heading of zero degrees has a local
+    /// reference frame according to the diagram below.
+    ///
+    ///       Y
+    ///       ^
+    ///       |
+    ///       |
+    ///       O--->X(forward)
```
Comment (azeey) [inline include/gz/math/MecanumDriveOdometry.hh:48]: Can you mention the caveat about wheel radius here? Most users won't see the TODO in the `.cc` file.

**Review by azeey — APPROVED** [review]
I haven't verified with gz-sim, but I've looked at the equations and the referenced [website](https://robohub.org/drive-kinematics-skid-steer-and-mecanum-ros-twist-included/) and it all looks good to me. I'd say the assumption of a `1.0` wheel radius should be explicitly stated in the class header and in the gz-sim system that uses this.

File: include/gz/math/MecanumDriveOdometry.hh (hunk 7 above)
Comment (mjcarroll) [reply include/gz/math/MecanumDriveOdometry.hh]: Went ahead and did it.

File: include/gz/math/MecanumDriveOdometry.hh (hunk 8 above)
Comment (mjcarroll) [reply include/gz/math/MecanumDriveOdometry.hh:51]: Done.

**Review by azeey — APPROVED** [review]

**Pull request merged**